core/trino-main/src/main/java/io/trino/operator/scalar/CharacterStringCasts.java

```diff
@@ -111,7 +111,17 @@ public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long
             return Slices.allocate(toIntExact(y));
         }
 
-        codePoints.set(codePoints.size() - 1, codePoints.get(codePoints.size() - 1) - 1);
+        int lastCodePoint = codePoints.get(codePoints.size() - 1) - 1;
+        /*
+         * UTF-8 reserve codepoints from 0xD800 to 0xDFFF for encoding UTF-16
```

Comment:
That's actually not standard UTF-8. There's a variant that does this called CESU-8 (see https://en.wikipedia.org/wiki/CESU-8). Note that Trino is expected to deal with proper UTF-8, so it's not clear to me that we actually want this.

Reply:
That is the whole point. Now there it is possible to get an incorrect UTF8 string from the saturated floor cast of the correct one.
`saturatedFllorCast(2, "0xE000") == "0xDFFF..."`

Reply:
That would not be an incorrect UTF-8 string. It would be an incorrect CESU-8 string

Reply:
Yet it will produce an InvalidCodePointException. Floor cast should not do that.

Reply:
That's a different problem. Where is that error being thrown? We should fix that issue, instead.

Reply:
Airlift UTF-8 utils will throw exception if the string contains characters like that.

Reply:
I misunderstood the purpose of this patch. The proposed fix is correct